File: ios/GitaVani/GitaVani/State/ReadingProgress.swift
import Foundation

@Observable
class ReadingProgress {
    var lastReadChapter: Int {
        didSet { UserDefaults.standard.set(lastReadChapter, forKey: "lastReadChapter") }
    }

    var lastReadVerse: Int {
        didSet { UserDefaults.standard.set(lastReadVerse, forKey: "lastReadVerse") }
    }

    var hasProgress: Bool {
        lastReadChapter > 0 && lastReadVerse > 0
    }

    init() {
        let defaults = UserDefaults.standard
        lastReadChapter = defaults.integer(forKey: "lastReadChapter")
        lastReadVerse = defaults.integer(forKey: "lastReadVerse")
    }

    func update(chapter: Int, verse: Int) {
        lastReadChapter = chapter
        lastReadVerse = verse
    }
}

File: android/GitaVani/app/src/main/java/com/nikhilsi/gitavani/state/ReadingProgress.kt
package com.nikhilsi.gitavani.state

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class ReadingProgress(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("gitavani_progress", Context.MODE_PRIVATE)

    private val _lastReadChapter = MutableStateFlow(prefs.getInt("lastReadChapter", 0))
    val lastReadChapter: StateFlow<Int> = _lastReadChapter.asStateFlow()

    private val _lastReadVerse = MutableStateFlow(prefs.getInt("lastReadVerse", 0))
    val lastReadVerse: StateFlow<Int> = _lastReadVerse.asStateFlow()

    val hasProgress: Boolean
        get() = _lastReadChapter.value > 0 && _lastReadVerse.value > 0

    fun update(chapter: Int, verse: Int) {
        _lastReadChapter.value = chapter
        _lastReadVerse.value = verse
        prefs.edit()
            .putInt("lastReadChapter", chapter)
            .putInt("lastReadVerse", verse)
            .apply()
    }
}
